Convert DynamicStressDivergenceTensors to AD

sveerara opened the pull request:
- Converts DynamicStressDivergenceTensors to AD. 
- Converting InertialForce to AD requires some more work because functions such as adUDotDot, adDuDotDu etc have to be added and I am not sure if any changes are required in the time integrators. So InertialForce is not converted in this PR.
- ADComputeFiniteStrainElasticStress serves as the parent class for one of the materials in MASTODON, so a line is added to the end of the ADComputeFiniteStrainElasticStress to enable this. 

refs #12650

sveerara committed:
Convert Dynamic StressDivergenceTensors to AD #12650

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/349870/) on f964ce9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13545/site)

This comment will be updated on new commits.

sveerara commented:
@dschwen, can you please review this PR? Thanks!

dschwen commented on modules/tensor_mechanics/include/kernels/ADDynamicStressDivergenceTensors.h:
```diff
@@ -0,0 +1,44 @@
+  const MaterialProperty<RankTwoTensor> & _stress_old;
+
+  // Rayleigh damping parameter _zeta and HHT time integration parameter _alpha
+  const MaterialProperty<Real> & _zeta;
```
If `_zeta` is expected to depend on non-linear variables this line would have to be 

```
const ADMaterialProperty(Real) & _zeta;
```

sveerara replied:
Thanks for comment, @dschwen. _zeta is an input parameter that can be provided by the user as a function of space but it does not depend on the non-linear variables. So I have not made this change.

dschwen commented on modules/tensor_mechanics/src/materials/ADComputeFiniteStrainElasticStress.C:
```diff
@@ -60,3 +60,6 @@ ADComputeFiniteStrainElasticStress<compute_stage>::computeQpStress()
 }
+
+// explicit instantiation is required for AD base classes
+adBaseClass(ADComputeFiniteStrainElasticStress);
```
This should not be necessary, as `ADComputeFiniteStrainElasticStress` gets registered and thus instantiated automatically.

sveerara replied:
@dschwen, when I comment out the above line and compile MASTODON, i get the following error:`dyld: Symbol not found: __ZN34ADComputeFiniteStrainElasticStressIL12ComputeStage0EE12initialSetupEv`  in all MASTODON tests. I have tried makeclean in both MOOSE and MASTODON and that doesn't seem to help. The MASTODON material model is in https://github.com/sveerara/mastodon/tree/ad_isoil

Any suggestions on what is causing this error?

lindsayad commented on modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests:
```diff
@@ -34,6 +34,19 @@
                    " numerical damping resulting from Hilber-Hughes-Taylor (HHT) time integration."
   [../]
+  [./rayleigh_hht_ad]
+    type = 'Exodiff'
```
When doing automatic differentiation work. I generally require that a `PetscJacobianTester` test be added

sveerara replied:
I have added a PetscJacobianTester test.

dschwen commented:
This is very confusing.

On Fri, Jun 14, 2019, 3:54 PM sveerara <notifications@github.com> wrote:

> *@sveerara* commented on this pull request.
> ------------------------------
>
> In
> modules/tensor_mechanics/src/materials/ADComputeFiniteStrainElasticStress.C
> <https://github.com/idaholab/moose/pull/13545#discussion_r294006075>:
>
> > @@ -60,3 +60,6 @@ ADComputeFiniteStrainElasticStress<compute_stage>::computeQpStress()
>    // Assign value for elastic strain, which is equal to the mechanical strain
>    _elastic_strain[_qp] = _mechanical_strain[_qp];
>  }
> +
> +// explicit instantiation is required for AD base classes
> +adBaseClass(ADComputeFiniteStrainElasticStress);
>
> @dschwen <https://github.com/dschwen>, when I comment out the above line
> and compile MASTODON, i get the following error:dyld: Symbol not found:
> __ZN34ADComputeFiniteStrainElasticStressIL12ComputeStage0EE12initialSetupEv
> in all MASTODON tests. I have tried makeclean in both MOOSE and MASTODON
> and that doesn't seem to help. The MASTODON material model is in
> https://github.com/sveerara/mastodon/tree/ad_isoil
>
> Any suggestions on what is causing this error?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13545?email_source=notifications&email_token=AABRMPT2B6YCWHHKHL7JJYTP2QHRJA5CNFSM4HXDZ5VKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3UOBLY#discussion_r294006075>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABRMPV7I2LJTXEE2QFWPM3P2QHRJANCNFSM4HXDZ5VA>
> .
>

lindsayad commented:
> This is very confusing.

For implicit instantiations the compiler will only generate symbols for the methods that are used. In registration, that `initialSetup` method is not instantiated so no symbols will be generated for it. However, the explicit instantiation of the class template will generate symbols for all definitions visible in the translation unit ***ahead of the explicit instantiation*** (note that at least on my Mac, `g++` just requires that it be visible in the translation unit, but `clang++` needs the definition ahead of the explicit instantiation). So we  need those lines at the bottom.

sveerara commented:
@lindsayad and @dschwen, thanks for looking into this issue. 

I did not remove the line at the end of ADComputeFiniteStrainElasticStress

permcody commented:
@sveerara - We forced in a few commits that failed clang format this weekend. I'm not sure if that's related to the current pre-check failure you are seeing or not, but I bet it is. Can you pull current "devel" and rebase this PR? That should get you past this odd problem.

sveerara commented:
@permcody, that fixed the clang-format test. Thanks!

dschwen commented on modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md:
```diff
@@ -0,0 +1,13 @@
+
+## Description
+
+This class computes the stress divergence and the stiffness proportional Rayleigh damping for structural dynamics problems. Each ADDynamicStressDivergenceTensors input block computes force in one direction. So, a separate ADDynamicStressDivergenceTensors input block should be set up for each coordinate direction. The [DynamicTensorMechanics](/DynamicTensorMechanicsAction.md) action automatically  ...
```
```suggestion
This class computes the stress divergence and the stiffness proportional Rayleigh damping for structural dynamics problems. Each ADDynamicStressDivergenceTensors input block computes force in one direction. So, a separate ADDynamicStressDivergenceTensors input block should be set up for each coordinate direction. The [DynamicTensorMechanics](/DynamicTensorMechanicsAction.md) action automatically sets up the ADDynamicStressDivergenceTensors input block in all coordinate directions. More information about the residual calculation and usage can be found at [Dynamics](Dynamics.md).
```

dschwen replied:
P.S.: This is one reason I prefer hard wrapped .md files (and .tex files). Spot the change!

lindsayad replied:
What do you mean by hard wrapped? An enforced line length?

dschwen replied:
Sorry that was a typo, I meant "hard rapped".

dschwen replied:
Yeah, of course, hard line breaks before the line length exceeds 80 chars. Most editors have a reflow hotkey, don't they?

lindsayad replied:
> Sorry that was a typo, I meant "hard rapped".

You like hard rap? 😄 

> line length exceeds 80 chars

80? Is that a standard of some kind? I personally use 100 for consistency with our clang-format, and yes I set that with my editor.

dschwen replied:
https://hackernoon.com/does-column-width-of-80-make-sense-in-2018-50c161fbdcf6

80 chars is an ancient convention! And you can debate (see above) whether it makes sense for code with wider screens. But I think the wider screens argument is misguided. Just use multiple editor views side by side. Text with longer lines is simply less readable, regardless of screen width:

https://practicaltypography.com/line-length.html

dschwen commented on modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C:
```diff
@@ -0,0 +1,91 @@
+    ADDynamicStressDivergenceTensors,
+    ADStressDivergenceTensors,
+    params.addClassDescription(
+        "Residual due to stress related Rayleigh damping and HHT time integration terms ");
```
```suggestion
        "Residual due to stress related Rayleigh damping and HHT time integration terms");
```

dschwen commented on modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C:
```diff
@@ -0,0 +1,91 @@
+   *_alpha*_zeta/dt Div sigma_older
+   */
+
+  ADResidual residual = 0.0;
```
```suggestion
  ADResidual residual;
```

dschwen commented on modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C:
```diff
@@ -0,0 +1,91 @@
+  {
+    // If static inialization is true, then in the first step residual is only Ku which is
+    // stress.grad(test).
+    residual += _stress[_qp].row(_component) * _grad_test[_i][_qp];
```
```suggestion
    residual = _stress[_qp].row(_component) * _grad_test[_i][_qp];
```

dschwen commented on modules/tensor_mechanics/src/kernels/ADDynamicStressDivergenceTensors.C:
```diff
@@ -0,0 +1,91 @@
+  }
+  else if (_dt > 0)
+  {
+    residual +=
```
```suggestion
    residual =
```

dschwen replied:
This looks petty, but an assignment might actually be faster than an unnecessary addition to a zero dual number.

dschwen commented on modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests:
```diff
@@ -34,6 +34,31 @@
+    input = 'wave_rayleigh_hht_AD.i'
+    exodiff = 'wave_rayleigh_hht_out.e'
+    abs_zero = 1e-09
+    compiler = 'GCC CLANG'
```
do we still need this restriction? We dropped intel compiler support a few months ago.

dschwen commented on modules/tensor_mechanics/test/tests/dynamics/wave_1D/tests:
```diff
@@ -34,6 +34,31 @@
+                 " both Rayleigh damping and numerical damping resulting from"
+                 " Hilber-Hughes-Taylor (HHT) time integration when automatic"
+                 " differentiation is used."
+    prereq = "rayleigh_hht_ad"
```
The pattern in many tests now is to suppress outputs in the jacobian test. Then you should be able to run it concurrently with the exidiff test.

lindsayad replied:
We actually explicit set `Outputs/exodus=false` by default with the `PetscJacobianTester` now. You can set `turn_off_exodus_output=True` in the test spec if you do want the exodus output. This way of turning off exodus output from the command line would not save you if you have an exodus sub-block in the Outputs block, however.

dschwen reviewed (changes requested):
Your `InertialForce` kernel depends on velocity and acceleration, which are AuxVariables that have no derivatives (and probably shouldn't because they are previous' time step values). Is this a lagging scheme? (just my curiosity)

lindsayad commented:
git blame aside would you argue for a change in our .clang-format file? I’m guessing there was a debate about this at the inception of clang-format. 

> On Jun 20, 2019, at 11:04 AM, Daniel Schwen <notifications@github.com> wrote:
> 
> @dschwen commented on this pull request.
> 
> In modules/tensor_mechanics/doc/content/source/kernels/ADDynamicStressDivergenceTensors.md:
> 
> > @@ -0,0 +1,13 @@
> +# ADDynamicStressDivergenceTensors
> +
> +!syntax description /Kernels/ADDynamicStressDivergenceTensors
> +
> +## Description
> +
> +This class computes the stress divergence and the stiffness proportional Rayleigh damping for structural dynamics problems. Each ADDynamicStressDivergenceTensors input block computes force in one direction. So, a separate ADDynamicStressDivergenceTensors input block should be set up for each coordinate direction. The [DynamicTensorMechanics](/DynamicTensorMechanicsAction.md) action automatically sets up the ADDynamicStressDivergenceTensors input block in all coordinate direction. More information about the residual calculation and usage can be found at [Dynamics](Dynamics.md).
> https://hackernoon.com/does-column-width-of-80-make-sense-in-2018-50c161fbdcf6
> 
> 80 chars is an ancient convention! And you can debate (see above) whether it makes sense for code with wider screens. But I think the wider screens argument is misguided. Just use multiple editor views side by side. Text with longer lines is simply less readable, regardless of screen width:
> 
> https://practicaltypography.com/line-length.html
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

dschwen commented:
Hm, I'm not advocating to change our code style standard, I'd just like to have some formatting convention for md files. Those line lengths don't necessarily have to be the same.

sveerara committed:
Addressing comments on ADDynamicStressDivergence #12650

dschwen reviewed (dismissed):
👍

sveerara commented:
@dschwen, thanks for your comments on the PR. I have addressed your comments. 

Regarding your comment on InertialForce kernel, it currently works both using velocity/acceleration auxvariables and also using the derivatives computed by the timeintegrator (if the auxvariables are not provided as input). The aim is to deprecate the aux variables version once all the necessary support objects (such as postprocessors to print out first or second time derivatives) are available for the time integrator version.

dschwen commented:
Uh oh, the failure is my fault! I didn't pay close enough attention and missed that there is a code path that leaves `residual` uninitialized! Sorry, but add an `else residual = 0.0;` before the return.

sveerara commented:
@dschwen, thanks for catching the error. I should have compiled it on my laptop before updating the PR. All tests are passing now.
